File: src/knapsack/mip-knapsack.jl
using JuMP
using HiGHS

function knapsack(W, w, v)
	model = Model(HiGHS.Optimizer)
    set_silent(model)
	n = length(w)

	@variable(model, items[1:n], Bin)
	@objective(model, Max, sum(v[i]*items[i] for i in 1:n))
	@constraint(model, sum(w[i]*items[i] for i in 1:n) <= W)

	optimize!(model)
	return JuMP.value.(items)
end

getitems(items) = findall(x -> x > 0.5, items)

function readwv(filename)
	lines = open(readlines, filename)
	W = parse(Int64, match(r"\d+", lines[1]).match)
	N = parse(Int64, match(r"\d+", lines[2]).match)
	
	w = fill(0, N)
	v = fill(0, N)
	
	for (i,line) in enumerate(lines[3:end])
		matches = collect(eachmatch(r"\d+", line))
		w[i] = parse(Int64, matches[1].match)
		v[i] = parse(Int64, matches[2].match)
	end

    return (W, w, v)
end

W, w, v = readwv(ARGS[1])
@time items = knapsack(W,w,v)
print(getitems(items))

File: src/knapsack/mip-knapsack.py
import sys
import re
import numpy as np
import gurobipy as grb

filename = sys.argv[1]

f = open(filename, "r")
lines = f.read().split("\n")[:-1]
f.close()

W = int(re.search(r"\d+", lines[0]).group(0))

wtsvals = [list(map(int,re.findall(r"\d+", line))) for line in lines[2:]]
w = np.zeros(len(wtsvals))
v = np.zeros(len(wtsvals))
for i, (wt, vl) in enumerate(wtsvals):
    w[i] = wt
    v[i] = vl

def knapsack(W, w, v):
    n = len(w)
    model = grb.Model()
    item = model.addVars(n, vtype=grb.GRB.BINARY)
    model.setObjective(grb.quicksum(v[i]*item[i] for i in range(n)), grb.GRB.MAXIMIZE)
    model.addConstr(grb.quicksum(w[i]*item[i] for i in range(n)) <= W)

    model.optimize()

    return [i+1 for i in range(n) if item[i].X > 0.5]

print(knapsack(W, w, v))

